EDU-131: What is a Logger? (#2255)
temporalio/documentation: docs-src/concepts/what-is-a-logger.md

Reviewer: Arguably the concept of a "logger" is a general programming one and not specific to Temporal nor does Temporal have enough non-SDK-specific things to say about a logger to deserve its own page IMO.

---
id: what-is-a-logger
title: What is a Logger?
sidebar_label: Logger
description: A Logger provides an API containing the logic for recording runtime behavior.
tags:
  - term
  - explanation
  - logger
  - api
  - sdks
---

A logger provides an API containing the logic needed to record details about the application behavior at runtime.
The output posts timestamps alongside the logged values, making it easier to evaluate the state and flow of [Workflows](/concepts/what-is-a-workflow) and [Activities](/concepts/what-is-an-activity).

[Temporal SDKs](/concepts/what-is-a-temporal-sdk) provide both a logging API and a built-in logger implementation.

> The logging implementation provided by the SDK emits messages to the standard output of the [Worker](/concepts/what-is-a-worker) executing the code.

Reviewer: This is not true of all SDKs. Many just use existing APIs and default to no logs.

Reviewer 2: @cretz Is there a place where the details are documented, at least internally? 

I'm most familiar with the Go SDK, which works as described. The Java SDK provides a [replay-aware logger](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/internal/logging/ReplayAwareLogger.html), which the API docs suggest is accessible via `workflow.getLogger`, although there is no Activity logger. My understanding is that the TypeScript SDK recently added a logger. I don't know about the others. 

Perhaps since there is so much variation in the logging support offered by each SDK, documenting how it works might be better done as part of each SDK-specific dev guide.

Reviewer: > Is there a place where the details are documented, at least internally?
Each SDK is different here. I am not sure we document them outside of SDK reference (even if we do there). I'd totally support documenting the SDK-specific details. That's unrelated to a general page.

> I'm most familiar with the Go SDK, which works as described. The Java SDK provides a [replay-aware logger](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/internal/logging/ReplayAwareLogger.html), which the API docs suggest is accessible via workflow.getLogger, although there is no Activity logger. My understanding is that the TypeScript SDK recently added a logger. I don't know about the others.

This comment is very specifically on the section stating that the implementation provided by the SDK emits to standard output. Not only is the standard output part not necessarily true. Python and .NET and Java all use ecosystem native logging that may not log anything by default (e.g. slf4j I think does nothing unless you include a JAR that has an impl).

> Perhaps since there is so much variation in the logging support offered by each SDK, documenting what is available and how to use it might be better done as part of each SDK-specific dev guide.

Exactly! I made this comment just below at https://github.com/temporalio/documentation/pull/2255#discussion_r1280933635

The Temporal logging API automatically injects helpful information (such as [Workflow ID](/concepts/what-is-a-workflow-id)) about the execution into the output.

> In addition, the logging API is replay-aware, preventing the duplication of log messages if History Replay is used to restore the app after a crash.

Reviewer: ```suggestion
In addition, the workflow logging API is replay-aware, preventing the duplication of log messages during replay.
```

Doesn't apply to activities and replay is not always about crashes.

> Although the output is timestamped and given a log level (similar to other logging APIs out there), the SDK does not support output customization.

Reviewer: > timestamped

This is SDK specific

> the SDK does not support output customization

This is not true. Most logging interfaces let you implement whatever approach you want. And ones using built-in logging frameworks (Java slf4j, Python logging module, .NET logging extension, etc) all support _lots_ of output customization. SDK just uses these like any other library would, there's nothing Temporal specific here in many cases (except for the extra properties we add and the skip-during-replay we offer).

Reviewer: > I filed the original issue because a user was confused about where the output from the Workflow and Activity loggers in the Go SDK went and how to change its format and/or destination.

But this doc we're reviewing is not for Go SDK

> AFAIK, there is no way for a developer to customize the format or output destination unless they write their own implementation (or an adapter class for an existing third-party package like zap), yet I was unable to find that documented anywhere.

Yes, it depends on the implementation they have. We provide an interface, how that interface is implemented is up to users. Formatting and output destination are but a couple of the many options one may choose to think about when implementing. We should move zap into contrib (https://github.com/temporalio/sdk-go/issues/704) as well as the upcoming `slog` (https://github.com/temporalio/sdk-go/issues/1158). But this PR doesn't help you solve your original problem of how to customize format or output destination.

> For an SDK like Java, which appears to use SLF4J, that might just be a couple of sentences that say

Right, see https://docs.temporal.io/dev-guide/java/observability#logging. That can be expanded/improved.

> Documentation for other SDKs could go into more detail, where necessary. To be clear, I don't think we need to cover how to configure a third-party logging package, but IMHO we should at least identify which one is being used by the SDK's logger (where applicable) so that people can find the docs provided by that project.

:+1:

Reviewer 2: > Right, see https://docs.temporal.io/dev-guide/java/observability#logging. That can be expanded/improved.

I hadn't seen this before, but it's pretty much what I had in mind when I filed the issue (although I didn't realize how different each SDK was).

> If you need to customize logger output, replace the logging implementation with an adapter that maps Temporal's API calls to a third-party library's implementation.

Reviewer: This is not true across SDKs. If you need to customize logger output in Python for example, you configure it, you don't replace anything.

:::caution
It's permissible to use an API provided by a third-party logging library in your Workflow and Activity Definitions.

> However, third-party APIs won't automatically inject information, and may duplicate logs upon a crash.

Reviewer: This is not necessarily true. We often wrap the third party libraries, not even invoking them on replay. And remember, replay situations are for more than just crashes.

:::

> For most cases, Temporal's API and default logging implementations will be enough for monitoring execution state.

Reviewer: I don't think this is true in all cases. For Go and Python and .NET I think the defaults are usually _not enough_.

Reviewer 3: Would go for something more middle ground, like `defaults may be enough for some, while others may require customizations...`. In most cases, those who needs more than the default probably already know that they need more.

Reviewer: > defaults may be enough for some, while others may require customizations

That's so obvious of a statement is it worth saying?

> If you require more control over the format, threshold, and destination of output, consider using a third-party implementation alongside Temporal's logging API, or integrate a third-party library.

Reviewer: This is not true. You don't always need third party. Many times the full advanced logging implementation is built in to the standard library (e.g. Python and the upcoming Go slog).
